File: bindings/python/tests/MFrontGenericBehaviourInterfaceTest3.py
# -*- coding: utf-8 -*-

import os
try:
    import unittest2 as unittest
except ImportError:
    import unittest
import mgis.behaviour as mgis_bv


class IntegrateTest(unittest.TestCase):

    def test_pass(self):

        stress_options = mgis_bv.FiniteStrainBehaviourOptionsStressMeasure
        to_options = mgis_bv.FiniteStrainBehaviourOptionsTangentOperator
        # path to the test library
        lib = os.environ['MGIS_TEST_BEHAVIOURS_LIBRARY']
        version = os.environ['MGIS_TEST_TFEL_VERSION']
        h = mgis_bv.Hypothesis.Tridimensional
        o = mgis_bv.FiniteStrainBehaviourOptions()
        o.stress_measure = stress_options.PK1
        o.tangent_operator = to_options.DPK1_DF
        b = mgis_bv.load(o, lib, 'FiniteStrainSingleCrystal', h)
        self.assertTrue(b.tfel_version == version, "invalid TFEL version")
        self.assertTrue(len(b.gradients) == 1, 'invalid number of gradients')
        self.assertTrue(b.gradients[0].name == 'DeformationGradient',
                        'invalid gradient name')
        self.assertTrue(b.gradients[0].getType() == 'Tensor',
                        'invalid gradient type')
        self.assertTrue(b.gradients[0].type == mgis_bv.VariableType.TENSOR,
                        'invalid gradient type')
        self.assertTrue(b.gradients[0].type == mgis_bv.VariableType.Tensor,
                        'invalid gradient type')
        self.assertTrue(len(b.thermodynamic_forces) == 1,
                        'invalid number of thermodynamic_forces')
        pk1 = b.thermodynamic_forces[0]
        self.assertTrue(pk1.name == 'FirstPiolaKirchhoffStress',
                        'invalid thermodynamic force  name')
        self.assertTrue(pk1.getType() == 'Tensor',
                        'invalid thermodynamic force  type')
        self.assertTrue(pk1.type == mgis_bv.VariableType.TENSOR,
                        'invalid thermodynamic force  type')


if __name__ == '__main__':
    unittest.main()

File: tests/MFrontGenericBehaviourInterfaceTest3.cxx
/*!
 * \file   MFrontGenericBehaviourInterfaceTest3.cxx
 * \brief
 * \author Thomas Helfer
 * \date   14/05/2019
 * \copyright (C) Copyright Thomas Helfer 2018.
 * Use, modification and distribution are subject
 * to one of the following licences:
 * - GNU Lesser General Public License (LGPL), Version 3.0. (See accompanying
 *   file LGPL-3.0.txt)
 * - CECILL-C,  Version 1.0 (See accompanying files
 *   CeCILL-C_V1-en.txt and CeCILL-C_V1-fr.txt).
 */

#include <iostream>
#include <cstdlib>
#include "MGIS/Behaviour/Behaviour.hxx"

int main(const int argc, const char* const* argv) {
  using namespace mgis::behaviour;
  constexpr const auto h = Hypothesis::TRIDIMENSIONAL;
  bool success = true;
  auto check = [&success](const bool b, const std::string& msg) {
    if (!b) {
      success = false;
      std::cerr << msg << '\n';
    }
    return b;
  };
  if (!check(argc == 3, "expected three arguments")) {
    return EXIT_FAILURE;
  }
  try{
    auto o = FiniteStrainBehaviourOptions{};
    o.stress_measure = FiniteStrainBehaviourOptions::PK1;
    o.tangent_operator = FiniteStrainBehaviourOptions::DPK1_DF;
    const auto d = load(o, argv[1], argv[2], h);
    check(d.symmetry == Behaviour::ORTHOTROPIC, "invalid behaviour symmetry");
    if (check(d.gradients.size() == 1u, "invalid number of gradients")) {
      check(d.gradients[0].name == "DeformationGradient", "invalid gradient name");
      check(d.gradients[0].type == Variable::TENSOR, "invalid gradient type");
    }
    if (check(d.thermodynamic_forces.size() == 1u,
	      "invalid number of thermodynamic_forces")) {
      check(d.thermodynamic_forces[0].name == "FirstPiolaKirchhoffStress",
	    "invalid flux name");
      check(d.thermodynamic_forces[0].type == Variable::TENSOR,
	    "invalid flux type");
    }
  } catch(std::exception& e){
    std::cerr << e.what() << std::endl;
    return EXIT_FAILURE;
  }
  return success ? EXIT_SUCCESS : EXIT_FAILURE;
}
